src/python/pants/rules/core/cloc.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pathlib import PurePath
+from typing import Set
+
+from pants.backend.graph_info.subsystems.cloc_binary import ClocBinary
+from pants.base.build_environment import get_buildroot
+from pants.base.specs import Specs
+from pants.engine.console import Console
+from pants.engine.fs import Digest, FilesContent, InputFileContent
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import HydratedTargets, TransitiveHydratedTargets
+from pants.engine.rules import console_rule, optionable_rule
+from pants.engine.selectors import Get
+
+
+class CountLinesOfCode(LineOriented, Goal):
+  name = 'cloc'
+
+  @classmethod
+  def register_options(cls, register) -> None:
+    super().register_options(register)
+    register('--transitive', type=bool, fingerprint=True, default=True,
+             help='Operate on the transitive dependencies of the specified targets.  '
+                  'Unset to operate only on the specified targets.')
+    register('--ignored', type=bool, fingerprint=True,
+             help='Show information about files ignored by cloc.')
+
+
+@console_rule(CountLinesOfCode, [Console, CountLinesOfCode.Options, ClocBinary, Specs])
+def run_cloc(console, options, cloc_binary, specs):
+  """Runs the cloc perl script in an isolated process"""
+
+  cloc_binary_path = cloc_binary.select()
+
+  transitive = options.values.transitive
+  ignored = options.values.ignored
+
+  if (transitive):
+    targets = yield Get(TransitiveHydratedTargets, Specs, specs)
+    all_target_adaptors = { t.adaptor for t in targets.roots }.union({ t.adaptor for t in targets.closure })
+  else:
+    targets = yield Get(HydratedTargets, Specs, specs)
+    all_target_adaptors = { t.adaptor for t in targets }
+
+  build_root = get_buildroot()
+  source_paths: Set[str] = set()
+  for t in all_target_adaptors:
+    sources = getattr(t, 'sources', None)
+    if sources is not None:
+      for file in sources.snapshot.files:
+        source_paths.add(str(PurePath(build_root, file)))
+
+  file_content = bytes('\n'.join(source_paths), 'utf-8')
```

Comment:
Preferred to convert via `.encode()`:

```suggestion
  file_content = '\n'.join(source_paths).encode()
```

I remember during the Py3 migration having some wonkiness with calling `bytes()`, plus this way you can leave off `utf-8`.